docs/src/writing_good_rules.md

```diff
@@ -229,6 +229,86 @@ Note that the function does not have to be $\mathbb{R} \rightarrow \mathbb{R}$.
 In fact, any number of scalar arguments is supported, as is returning a tuple of scalars.
 
 See docstrings for the comprehensive usage instructions.
+
+
+### Be careful about pullbacks closures calling other methods of themselves
+
+Due to [JuliaLang/Julia#40990](https://github.com/JuliaLang/julia/issues/40990), a closure calling another (or the same) method of itself often comes out uninferable (and thus effectively type-unstable).
+This can be avoided by moving the pullback outside the function.
+For example:
+
+```julia
+double_it(x::AbstractArray) = 2 .* x
+
+function ChainRulesCore.rrule(::typeof(double_it), x)
+    double_it_pullback(ȳ::AbstractArray) = (NoTangent(), 2 .* ȳ)
+    double_it_pullback(ȳ::AbstractThunk) = double_it_pullback(unthunk(ȳ))
+    return double_it(x), double_it_pullback
+end
+```
+Ends up infering a return type of `Any`
+```julia
+julia> _, pullback = rrule(double_it, [2.0, 3.0])
+([4.0, 6.0], var"#double_it_pullback#8"(Core.Box(var"#double_it_pullback#8"(#= circular reference @-2 =#))))
+
+julia> @code_warntype pullback(@thunk([10.0, 10.0]))
+Variables
+  #self#::var"#double_it_pullback#8"
+  ȳ::Core.Const(Thunk(var"#9#10"()))
+  double_it_pullback::Union{}
+
+Body::Any
+1 ─ %1 = Core.getfield(#self#, :double_it_pullback)::Core.Box
+│   %2 = Core.isdefined(%1, :contents)::Bool
+└──      goto #3 if not %2
+2 ─      goto #4
+3 ─      Core.NewvarNode(:(double_it_pullback))
+└──      double_it_pullback
+4 ┄ %7 = Core.getfield(%1, :contents)::Any
+│   %8 = Main.unthunk(ȳ)::Vector{Float64}
+│   %9 = (%7)(%8)::Any
+└──      return %9
+```
+
+This can be solved by moving the pullbacks outside the function so they are not closures, and thus to not run into this upstream issue.
+In this case that is fairly simple, since this example doesn't close over anything (if it did then would need a closure calling an outside function that calls itself).
```

Comment:
I wonder whether we should add an example for that. Probably we can just link to an existing pattern in the code, such as https://github.com/JuliaDiff/ChainRules.jl/blob/773039a2dc0a1938f61cf26012b1223c942bc18f/src/rulesets/LinearAlgebra/structured.jl#L107-L116?